### PR TITLE
#168 - Experimental JSON CAS support

### PR DESCRIPTION
**Pull request by reckart**
- [x] Basic JSON CAS support
- [x] Randomized tests

### CODE REVIEW

File: cassis/cas.py [hunk 1]
```diff
@@ -171,6 +177,40 @@ def __init__(self, typesystem: TypeSystem):
 class Cas:
     """A CAS object is a container for text (sofa) and annotations"""
 
+    NAME_SPACE_UIMA_CAS = "uima" + TypeSystem.NAMESPACE_SEPARATOR + "cas"
```
Comment (jcklie) [inline cassis/cas.py:180]: Are these not already in the typesystem.py ?

File: cassis/cas.py [hunk 2]
```diff
@@ -514,8 +573,36 @@ def to_xmi(self, path: Union[str, Path, None] = None, pretty_print: bool = False
...
+
+        Returns:
+            If `path` is None, then the JSON representation of this CAS is returned as a string
+
+        """
+        from cassis.json import CasJsonSerializer
+
+        return self.serialize(CasJsonSerializer(), path, pretty_print)
+
+    def serialize(self, serializer, path: Union[str, Path, None] = None, pretty_print: bool = False):
```
Comment (jcklie) [inline cassis/cas.py:594]: Does this need to be public? Also, a type or protocol for serializers would be nice

File: cassis/cas.py [hunk 3]
```diff
@@ -591,7 +678,10 @@ def _find_all_fs(self) -> Iterable[FeatureStructure]:
                         openlist.append(referenced_fs)
 
         # We do not want to return cas:NULL here as we handle serializing it later
-        all_fs.pop(0, None)
+        for fs_id, fs in list(all_fs.items()):
```
Comment (jcklie) [inline cassis/cas.py:681]: Why do you convert it to a list?

File: cassis/typesystem.py [hunk 4]
```diff
@@ -349,6 +349,8 @@ def descendants(self) -> Iterator["Type"]:
 
 
 class TypeSystem:
+    NAMESPACE_SEPARATOR = "."
```
Comment (jcklie) [inline cassis/typesystem.py:352]: Why is this here?

File: tests/test_files/json/fs_as_array/ser-ref/casWithSofaDataArray/data.json [hunk 5]
```diff
@@ -0,0 +1,21 @@
...
+  "%VIEWS" : {
+    "_InitialView" : {
+      "%SOFA" : 0,
+      "%INDEX" : [ ]
+    }
+  },
+  "%FEATURE_STRUCTURES" : [ {
+    "%ID" : 1,
+    "%TYPE" : "uima.cas.ByteArray",
+    "%ELEMENTS" : "VGhpcyBpcyBhIHRlc3Q="
```
Comment (jcklie) [inline tests/test_files/json/fs_as_array/ser-ref/casWithSofaDataArray/data.json:12]: Why is the type of a byte array a string? Is that how UIMA rolls or is it temporary?

File: tests/util.py [hunk 6]
```diff
@@ -39,6 +40,39 @@ def assert_xml_equal(actual: Union[IO, str], expected: Union[IO, str]):
...
+    """Checks whether the JSON trees behind `actual` and `expected` are equal.
+
+    Args:
+        actual: The actual JSON
+        expected: The expected JSON
+
+    Throws:
+        AssertionError when json(actual) != json(expected)
+    """
+    actual_json = json.dumps(json.loads(actual), sort_keys=True, indent=2)
```
Comment (jcklie) [inline tests/util.py:53]: Is that really the best way to compare two jsons? I would expect dict comparison to also work

Comment (reckart) [reply tests/util.py:53]: Well, it works and it produces a nicely diffable representation. No idea if it is the best way. What benefit would dict-comparison have?

File: tests/test_files/json/fs_as_array/ser-ref/casWithSofaDataArray/data.json (hunk 5 above)
Comment (reckart) [reply tests/test_files/json/fs_as_array/ser-ref/casWithSofaDataArray/data.json:12]: Not sure what you mean. If you mean that "elements" is a base64-encoded string, then I'd say it is simply what makes sense - it is also the default how Jackson serializes byte arrays into JSON.

File: cassis/typesystem.py (hunk 4 above)
Comment (reckart) [reply cassis/typesystem.py:352]: Because it is there in the Apache UIMA Java SDK.

File: cassis/cas.py (hunk 3 above)
Comment (reckart) [reply cassis/cas.py:681]: because I am modifying the dictionary as I iterate over it and don't want to get any concurrent modification problems.

File: cassis/cas.py (hunk 2 above)
Comment (reckart) [reply cassis/cas.py:594]: Doesn't need to be public. What is a protocol in this particular context? You mean like an interface?

File: cassis/cas.py (hunk 1 above)
Comment (reckart) [reply cassis/cas.py:180]: Not as constants, just as lists. Also, in the Apache UIMA Java SDK, these constants are in the Cas, so that's where I look for them ;)

File: tests/util.py (hunk 6 above)
Comment (jcklie) [reply tests/util.py:53]: It would be independent of the ordering. You can always add the diffable representation after the assert, e.g. `assert foo, "My error message`

File: tests/test_files/json/fs_as_array/ser-ref/casWithSofaDataArray/data.json (hunk 5 above)
Comment (jcklie) [reply tests/test_files/json/fs_as_array/ser-ref/casWithSofaDataArray/data.json:12]: I would have thought it is an array of ints < 255

File: cassis/typesystem.py (hunk 4 above)
Comment (jcklie) [reply cassis/typesystem.py:352]: But you seem to not use it

File: cassis/cas.py (hunk 2 above)
Comment (jcklie) [reply cassis/cas.py:594]: You can then let it start with an underscore

https://docs.python.org/3/library/typing.html#typing.Protocol

File: cassis/cas.py (hunk 1 above)
Comment (jcklie) [reply cassis/cas.py:180]: It would then be better to put them in the typesystem and let the lists use them there. For me they are more in the typesystem and not in the cas, as the former is where cassis uses them the most

File: tests/test_files/json/fs_as_array/ser-ref/casWithSofaDataArray/data.json (hunk 5 above)
Comment (reckart) [reply tests/test_files/json/fs_as_array/ser-ref/casWithSofaDataArray/data.json:12]: That is way less space-efficient.

File: tests/util.py (hunk 6 above)
Comment (reckart) [reply tests/util.py:53]: This is also independent of ordering because keys are sorted - but to be honest, I'd remove the key sorting because order actually matters. E.g. we want that the three main section have a specific order, also in an FS, the fields with the reserved names must come before the feature fields, etc.

File: cassis/typesystem.py (hunk 4 above)
Comment (reckart) [reply cassis/typesystem.py:352]: One of the constants defined in the Cas reference it.

File: tests/test_files/json/fs_as_array/ser-ref/casWithSofaDataArray/data.json (hunk 5 above)
Comment (jcklie) [reply tests/test_files/json/fs_as_array/ser-ref/casWithSofaDataArray/data.json:12]: How does UIMAj serialize it with XMI?

File: cassis/cas.py (hunk 1 above)
Comment (reckart) [reply cassis/cas.py:180]: Can the type system not import them from the CAS? I am be bit biased because I am used to the constants being in the CAS.

Comment (jcklie) [reply cassis/cas.py:180]: The CAS in cassis is mostly not using it, only the type system and serializer/deserializer do I think to I find it strange to put it in the cas.py where it is mostly not used and let others import it.

File: tests/test_files/json/fs_as_array/ser-ref/casWithSofaDataArray/data.json (hunk 5 above)
Comment (reckart) [reply tests/test_files/json/fs_as_array/ser-ref/casWithSofaDataArray/data.json:12]: https://github.com/apache/uima-uimaj/blob/main/uimaj-core/src/test/resources/CasSerializationDeserialization_XMI_1_0_PRETTY_Test/ser-ref/casWithSofaDataArray/data.xmi

```
<?xml version="1.0" encoding="UTF-8"?><xmi:XMI xmlns:tcas="http:///uima/tcas.ecore" xmlns:xmi="http://www.omg.org/XMI" xmlns:cas="http:///uima/cas.ecore" xmi:version="2.0">
    <cas:NULL xmi:id="0"/>
    <cas:Sofa xmi:id="2" sofaNum="1" sofaID="_InitialView" mimeType="text/plain" sofaArray="1"/>
    <cas:ByteArray xmi:id="1" elements="5468697320697320612074657374"/>
</xmi:XMI>
```

Comment (reckart) [reply tests/test_files/json/fs_as_array/ser-ref/casWithSofaDataArray/data.json:12]: But hex strings are old-skool, no?
